sdk/typespec/typespec_derive/src/lib.rs

```diff
@@ -0,0 +1,205 @@
+    pub fn new() -> Context {
+        Context { errors: Vec::new() }
+    }
+
+    pub fn emit_error(&mut self, span: Span, message: &str) {
+        let token_stream = quote::quote_spanned!(span=> compile_error!(#message));
+        self.errors.push(token_stream);
+    }
+
+    pub fn errors(&self) -> &[TokenStream] {
+        &self.errors
+    }
+}
+
+struct Attrs {
+    pub typespec_path: Option<Path>,
+    pub format: Option<Format>,
+}
+
+impl Attrs {
+    pub fn from_attrs(attrs: &[Attribute], ctx: &mut Context) -> Attrs {
+        let mut typespec_path = None;
+        let mut format = None;
+        for attr in attrs.iter().filter(|a| a.path().is_ident("typespec")) {
+            let Meta::List(meta_list) = &attr.meta else {
+                ctx.emit_error(attr.span(), "Invalid typespec attribute, expected attribute in form #[typespec(key = value)]");
+                continue;
+            };
+
+            meta_list.parse_nested_meta(|meta| {
+                let Some(ident) = meta.path.get_ident() else {
+                    ctx.emit_error(meta.path.span(), "Invalid typespec attribute, expected attribute in form #[typespec(key = value)]");
+                    return Ok(());
+                };
+                match ident.to_string().as_str() {
+                    "crate" => {
+                        let value = meta.value().unwrap();
+                        let Ok(value) = parse_literal_string(value) else {
+                            ctx.emit_error(value.span(), "Invalid value for 'crate' attribute, expected string literal");
+                            return Ok(());
+                        };
+                        let path = value.parse().unwrap();
+                        typespec_path = Some(path);
+                    }
+                    "format" => {
+                        let value = meta.value().unwrap();
+                        let Ok(lit) = parse_literal_string(value) else {
+                            ctx.emit_error(value.span(), "Invalid value for 'format' attribute, expected string literal");
+                            return Ok(());
+                        };
+                        format = Some(match lit.value().as_str() {
+                            "json" => Format::Json,
+                            "xml" => Format::Xml,
+                            x => {
+                                ctx.emit_error(lit.span(), &format!("Unknown format '{}'", x));
+                                return Ok(());
+                            }
+                        });
+                    }
+                    x => {
+                        ctx.emit_error(meta.path.span(), &format!("Unknown typespec attribute '#[typespec({})'", x));
+                    }
+                };
+                Ok(())
+            }).unwrap();
+        }
+        Attrs {
+            typespec_path,
+            format,
+        }
+    }
+}
+
+fn parse_literal_string(value: ParseStream) -> Result<LitStr, ()> {
```

Comment:
FWIW, helpers like this can and should stay in `lib.rs`. Same for `Context` above.